packages/wallet/ui/src/util/keyManagement.js

```diff
@@ -0,0 +1,429 @@
+ */
+const CosmosMessages = /** @type {const} */ ({
+  bank: {
+    MsgSend: {
+      typeUrl: '/cosmos.bank.v1beta1.MsgSend',
+    },
+  },
+  authz: {
+    MsgGrant: {
+      typeUrl: '/cosmos.authz.v1beta1.MsgGrant',
+    },
+    GenericAuthorization: {
+      typeUrl: '/cosmos.authz.v1beta1.GenericAuthorization',
+    },
+    MsgExec: {
+      typeUrl: '/cosmos.authz.v1beta1.MsgExec',
+    },
+  },
+  feegrant: {
+    MsgGrantAllowance: {
+      typeUrl: '/cosmos.feegrant.v1beta1.MsgGrantAllowance',
+    },
+    BasicAllowance: {
+      typeUrl: '/cosmos.feegrant.v1beta1.BasicAllowance',
+    },
+  },
+});
+
+/**
+ * `/agoric.swingset.XXX` matches package agoric.swingset in swingset/msgs.go
+ */
+export const SwingsetMsgs = /** @type {const} */ ({
+  MsgWalletAction: {
+    typeUrl: '/agoric.swingset.MsgWalletAction',
+    aminoType: 'swingset/WalletAction',
+  },
+  MsgWalletSpendAction: {
+    typeUrl: '/agoric.swingset.MsgWalletSpendAction',
+    aminoType: 'swingset/WalletAction',
+  },
+});
+
+/**
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   action: string,
+ * }} WalletAction
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   spendAction: string,
+ * }} WalletSpendAction
+ */
+
+export const SwingsetRegistry = new Registry([
+  ...defaultRegistryTypes,
+  // XXX should this list be "upstreamed" to @agoric/cosmic-proto?
+  [SwingsetMsgs.MsgWalletAction.typeUrl, MsgWalletAction],
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl, MsgWalletSpendAction],
+]);
+
+/**
+ * TODO: estimate fee? use 'auto' fee?
+ * https://github.com/Agoric/agoric-sdk/issues/5888
+ *
+ * @returns {import('@cosmjs/stargate').StdFee}
+ */
+export const zeroFee = () => {
+  const { coinMinimalDenom: denom } = stableCurrency;
+  const fee = {
+    amount: [{ amount: '0', denom }],
+    gas: '100000', // TODO: estimate gas?
+  };
+  return fee;
+};
+
+/** @type {import('@cosmjs/stargate').AminoConverters} */
+const SwingsetConverters = {
+  // TODO: #3628, #4654
+  // '/agoric.swingset.MsgProvision': {
+  //   /* ... */
+  // },
+  [SwingsetMsgs.MsgWalletAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ action, owner }) => ({
+      action,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ action, owner }) => ({
+      action,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+};
+
+// ack: justerest Sep 2019 https://stackoverflow.com/a/58110124/7963
```

Comment:
a little ugly but I see it's necessary. might want to reference `UNTIL https://github.com/microsoft/TypeScript/issues/16655` so we know when we can remove this

Reply:
oh. JS Arrays seem to have `flatMap` now... i.e. map and filter in one...